at-angular.ts

```diff
@@ -107,8 +151,36 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
+
+        return (targetClass: any): void => {
+
+            angular
+                .module(moduleName)
+                .config([
+                    '$provide',
+                    function($provide: angular.auto.IProvideService): void {
+
+                    $provide.decorator(targetProvider, [
+                       '$delegate',
+                        ($delegate: any): void => {
+
+                            $delegate[targetProvider] = $delegate;
+
+                            angular.extend($delegate, new targetClass());
```

Comment:
`angular.extend` (just like `Object.assign`) does shallow copy. It does not copy fields inherited from prototype. In case of TS/ES6 class instance all methods are bound to proto:

```typescript
class A {
  foo = 'a';
  bar() { return this.foo; }
}

let someInstance = new A();
let aDelegate = {baz: 'x'};
let extended = angular.extend(aDelegate, someInstance);

console.log(extended);
// => Object { foo: "a", baz: "x" }
```

methods will not be copied in such case.

You do not have this problem with object literals:

```typescript

let someInstance = {
  foo: 'a',
  bar() { return this.foo; }
}
let aDelegate = {baz: 'x'};
let extended = angular.extend(aDelegate, someInstance);

console.log(extended);
// => Object { foo: "a", baz: "x", bar: function() {…} }
```

That is why angular-generator did not work with `prototype`.